downloader/src/main/kotlin/VersionControlSystem.kt

```diff
@@ -283,11 +283,17 @@ abstract class VersionControlSystem {
             throw DownloadException("Unable to determine a revision to checkout.")
         }
 
-        var i = 0
-        val workingTreeRevision = revisionCandidates.find { revision ->
-            log.info { "Trying revision candidate '$revision' (${++i} of ${revisionCandidates.size})..." }
-            updateWorkingTree(workingTree, revision, pkg.vcsProcessed.path, recursive)
-        } ?: throw DownloadException("$type failed to download from URL '${pkg.vcsProcessed.url}'.")
+        val results = mutableListOf<Result<String>>()
+
+        revisionCandidates.forEachIndexed { index, revision ->
+            log.info { "Trying revision candidate '$revision' (${index + 1} of ${revisionCandidates.size})..." }
+            results += updateWorkingTree(workingTree, revision, pkg.vcsProcessed.path, recursive)
+            if (results.last().isSuccess) return@forEachIndexed
+        }
+
+        val workingTreeRevision = results.last().getOrElse {
```

Comment:
I believe that in case `results.last() is Failure == true` the handling is now missing and needs to be added.

Reply:
The "OrElse" part is what gets executed in that case.

Reply:
If result contains more than zero elements and all elements are of the type `Failure`, then, you say, `orElse` gets executed?

Reply:
Yes. But only the message of the last failure is included into the `DownloadException`.

What probably confuses you is that the `getOrElse` here is not the function on a collection, but the function on a `Result` object.

Reply:
And `results` cannot be empty as `revisionCandidates` is ensured to not be empty.

Reply:
> What probably confuses you is that the getOrElse here is not the function on a collection, but the function on a Result object.

Argh, yes that was exactly what was confusing me. Thanks for pointing it out.